spec/index.bs

```diff
@@ -1327,6 +1221,37 @@ Content-Security-Policy: trusted-types one two default
 </pre>
 </div>
 
+#### `trusted-types` Pre-Navigation check #### {#trusted-types-pre-navigation-check}
+
+Given a [[Fetch#concept-request|request]] (|request|), a string |navigation type|, two [=browsing contexts=]
+(|source| and |target|), and a [[CSP3#content-security-policy-object|policy]] (|policy|), this algorithm returns `"Blocked"`
+if a navigation violates the {{#trusted-types-directive|trusted-types}} directive's constraints and `"Allowed"`
+otherwise. This constitutes the {{#trusted-types-directive|trusted-types}} directive's [=pre-navigation check=]:
+
+Note: This algorithm assures that the code to be executed by a navigation to a `javascript:` URL will have to pass through a
+<a>default policy</a>'s `createScript` function, in addition to all other restrictions imposed by other CSP directives.
+
+1. If |request|'s [=request/url=]'s [=url/scheme=] is not `"javascript"`, return `"Allowed"` and abort further steps.
+1. Let |urlString| be the result of running the [=URL serializer=] on |request|'s [=request/url=].
+1. Let |encodedScriptSource| be the result of removing the leading `"javascript:"` from |urlString|.
+1. Let |defaultPolicy| be the result of executing [$Get default policy$] algorithm on |source|'s {{WindowProxy}}'s [=trusted type policy factory=].
+1. If |defaultPolicy| is `null`, return `"Blocked"` and abort further steps.
+1. Let |covertedScriptSource| be the result of executing [$Create a Trusted Type$] algorithm, with the following arguments:
+
+    * |defaultPolicy| as |policy|
+    * |encodedScriptSource| as |value|
+    * `"TrustedScript` as |trustedTypeName|
+    * &laquo; `"Location.href"` &raquo; as |arguments|
+1. If |convertedScriptSource| does not have a type {{TrustedScript}}, return `"Blocked"` and abort further steps.
+1. Set |urlString| to be the result of prepending `"javascript:"` to stringified |convertedScriptSource|.
+1. Let |newURL| be the result of running the [=URL parser=] on |urlString|. If the parser returns a failure, return `"Blocked"` and abort further steps.
+1. Set |request|'s [=request/url=] to |newURL|.
+
+     Note: No other CSP directives operate on `javascript:` URLs in a pre-navigation check. Other directives that check javascript: URLs
+           will operate on the modified URL later, in the [=inline check=].
```

Comment:
Do you want this to have the `eval()` behavior, wherein the policy's acceptance of the script by minting a `TrustedScript` allows it to bypass the absence of `unsafe-eval`? Or is it expected that a policy which disallows inline script will continue to disallow `javascript:` navigation?

Reply:
We could control this via [`trusted-script`](https://wicg.github.io/trusted-types/dist/spec/#trusted-script-csp-keyword) keyword in `script-src`. I.e. if there are scripty directives in your CSP, all should have `trusted-script`.

Reply:
I added the description on how that would behave with `'trusted-script'` CSP keyword. The possible complexity I see is that trusted-script depends the whole CSP lists (like it was the case for eval). I feel we can improve this later on, perhaps by requiring that 'trusted-script' is present on every CSP. This feels like better to be discussed in the CSP bugtracker later on, or in the ocntext of CSP next?